.devcontainer/Dockerfile

```diff
@@ -0,0 +1,10 @@
+FROM mcr.microsoft.com/devcontainers/python:3.11-bullseye@sha256:105bf6a63ab025206f019a371a735fec6553db0be520030c7a2fd0e002947232
+
+WORKDIR /app
+
+# Install Python dependencies
+COPY requirements.txt /app/
+RUN pip install -U 'pip>=20' && \
+    pip install --no-cache-dir --no-deps -r requirements.txt && \
+    pip install --no-cache-dir ipython && \
+    pip check --disable-pip-version-check
```

Comment:
I'm taking this as an opportunity to learn some of the Python stuff as well (hope you don't mind 😅 ).

I understand this is largely based on `docker/Dockerfile`, I think I grok most of the options here except the `--no-deps` one.

IIUC, this means [don't do dependency resolution](https://pip.pypa.io/en/stable/user_guide/#how-to-upgrade-and-migrate) -- because pip might throw some version icompatibility error when in reality the deps work together just fine?

Reply:
Packages on pypi can be changed in place, which could introduce a new dependency with code that we don't trust. For this reason it is sometimes considered good practice to set `--no-deps` to ensure no unverified dependencies are installed.

If a transitive dependency is missing from requirements.txt and requirements.txt contains hashes, then pip will refuse to install anything, because if one dependency has a hash set all dependencies must have a hash set. The most common reason I've seen for a missing transitive dependency is using different python versions, because packages can set version-specific dependencies, and pip-compile will only include transitive dependencies based on the current python version.

Another concrete reason to use `--no-deps` is that [pip has a bug with version resolution and extras](https://github.com/pypa/pip/issues/9644) that has been haunting me for years. if your requirements file contains hashes, and in your dependency tree you have something like `a` requires `c`, `b` requires `c[d]`, and your requirements file looks like:
```
a==1.0 --hash[...]
b==1.0 --hash[...]
c[d]==1.0 --hash[...]
```
then pip will fail because the requirements file doesn't contain a hash for `c` (without extras). using `--no-deps` fixes the issue. this impacts certain gcp sdk libraries that require something like either `sdk-common` or `sdk-common[protobuf]`.

And last but not least, because we expect requirements.txt to contain all transitive dependencies and pip-compile enforces this behavior, so `--no-deps` speeds things up because we don't need to check for deps.

Reply:
Thanks for the details, relud! I thought the reason for `--no-deps` was mainly the last thing you mentioned – we use pip-compile to resolve dependencies, so there is no reason to have pip resolve them _again_. But I remember that bug you mention, and it's definitely another good reason.

Installing IPython here is kind of a hack. If we ever want more dev-container-specific dependencies, we can figure out a better way of installing them.

Reply:
Thank you both for the information. So much behind one ostensibly simple option.

> Packages on pypi can be changed in place
Can you clarify what you mean here? E.g. if we specify `latest` or otherwise have not pinned a dependency at a specific version in `requirements.in`, and a package update is available?

> For this reason it is sometimes considered good practice to set --no-deps to ensure no unverified dependencies are installed.
What does unverified mean in this case?

Reply:
> Packages on pypi can be changed in place

if i have already published version 1.1.1, I can modify the packages hosted on pypi at any time, and the only indication will be different hash values. One reason why we have `requirements.in` and `requirements.txt` is so that we can use `pip-compile --generate-hashes` to add `--hash` arguments for each package in `requirements.txt`, so that if a package is modified we get an error instead of installing potentially malicious code.

>> set --no-deps to ensure no unverified dependencies are installed.
>
> What does unverified mean in this case?

I was using verified to mean "matches a package in `requirements.txt`".

I should have worded this better, because really it's having hashes in `requirements.txt` that provides most of the protection here. `--no-deps` would only be providing any protection if we didn't include hashes in `requirements.txt` and thus could install dependencies not present in `requirements.txt`, and then it would only protect against malicious code in those dependencies and not malicious code added to an existing dependency.

In theory before merging any dependabot PRs or adding any new dependencies directly, we could be checking the content of every package hash added to `requirements.txt` to make sure no malicious code has been included, and ~7 years ago i was advised to do that once by the data org python expert at the time. In practice I have not seen a project do this, and just using hashes is usually considered sufficient protection.